### ISSUE TITLE
fstrim: System hangs/breaks when fstrim runs (dual boot on SSD setup)

### ISSUE THREAD

**Issue by ahmetsait** [util-linux/util-linux#1929]
I'm having some issues with fstrim and I'm not entirely sure whether it's directly a bug in fstrim or some other parts of my system so I'm looking for some advice on how to debug/pinpoint this problem.

I have [a question on askubuntu](https://askubuntu.com/q/1442510/386600) where I described the issue with lots of details so I'm going to refrain from copying it all here, please take a look. I've also seen [a bug report on launchpad](https://bugs.launchpad.net/ubuntu/+source/systemd/+bug/1823721) that closely matches what I'm experiencing.

TLDR: When fstrim service runs (every week) I experience weird issues (apps freezing/breaking etc.) that forces me to emergency reboot. As I said, not sure if it's entirely fstrim's fault but posting here seemed like a good start.

**Comment by karelzak** [comment]
Well, poor performance during fstrim is something you can expect (this is the reason why it is usually scheduled for the night).  What SSD disk do you have? It seems some disks (for example, Samsung) perform really badly.

**Comment by ahmetsait** [comment]
Looks like I have an [INTEL SSDSCKHW240A4
](https://ark.intel.com/content/www/us/en/ark/products/75331/intel-ssd-530-series-240gb-2-5in-sata-6gbs-20nm-mlc.html).

**Comment by karelzak** [comment]
I guess the kernel is waiting for the device, but you can try `strace fstrim --listed-in /etc/fstab:/proc/self/mountinfo --verbose --quiet-unsupported` (I guess you have this in fstrim.service or in cron) to be sure.

IMHO this is something that we cannot fix in fstrim, maybe kernel developers will help.

**Comment by karelzak** [comment]
It's also possible that you have a lot of unused blocks on the SSD and it takes time to trim it.

**Comment by masatake** [comment]
@ahmetsait Just a question. Which file system do you use? XFS?

**Comment by ahmetsait** [comment]
@masatake I use ext4 and ntfs.

I was trying to debug this a bit more and I ran `fstrim --fstab --verbose | tee fstrim.log` and left the computer running at home to see if it locks up indefinitely. I've found it finished when I got back home, and according to the last modification time of `fstrim.log` it took about an hour:
```
/mnt/C: 37.1 GiB (39868157952 bytes) trimmed on /dev/sda3
/boot/efi: 60.8 MiB (63706112 bytes) trimmed on /dev/sda1
/: 570.6 MiB (598290432 bytes) trimmed on /dev/sda5
```
What I've also realized is that when fstrim calls `ioctl(3, FITRIM, {start=0, len=18446744073709551615, minlen=0}...` any application that tries to access `/mnt/C` hangs, and do not even respond to `killall -KILL` which makes me suspect there is some intricate issue with the ntfs-3g driver that cause these long freezes, since I expect to be able to use mounted partitions just fine while FITRIM command is in progress (otherwise it seriously shouldn't do it live).

Could you guys forward this to kernel / nfts-3g devs maybe?

**Comment by lczerner** [comment]
Hi @ahmetsait

ext4 developer and author of the FITRIM ioctl here. If I understand it correctly the problem is with the /mnt/C mount point. It is not entirely clear what file system is on that mount point, but you're talking about ntfs-3g driver. Is that what is on /mnt/C ? Are you sure it is a fuse driver ntfs-3g and not the upstream in-kernel ntfs3 driver? Can you provide the output of 'mount' ?

Looking at the kernel ntfs3 driver, even though I am not very familiar with the code, it looks like they are holding a semaphore for reading during the entire operation. This semaphore seems to be global fs-wide lock that is used on many places but it is also acquired for write during allocation. That would mean that at least new writes to the fs would block until the fstrim finishes on the file system. It looks like this is what you're experiencing.

If you're indeed using the kernel ntfs3 driver and that's the fs you're experiencing the problems with, please contact the authors directly (https://www.paragon-software.com/) or use their mailing list ntfs3@lists.linux.dev.

Thanks!
-Lukas

**Comment by ahmetsait** [comment]
@lczerner Output of `mount`:
```
sysfs on /sys type sysfs (rw,nosuid,nodev,noexec,relatime)
proc on /proc type proc (rw,nosuid,nodev,noexec,relatime)
udev on /dev type devtmpfs (rw,nosuid,relatime,size=3888916k,nr_inodes=972229,mode=755,inode64)
devpts on /dev/pts type devpts (rw,nosuid,noexec,relatime,gid=5,mode=620,ptmxmode=000)
tmpfs on /run type tmpfs (rw,nosuid,nodev,noexec,relatime,size=794920k,mode=755,inode64)
/dev/sda5 on / type ext4 (rw,relatime)
securityfs on /sys/kernel/security type securityfs (rw,nosuid,nodev,noexec,relatime)
tmpfs on /dev/shm type tmpfs (rw,nosuid,nodev,inode64)
tmpfs on /run/lock type tmpfs (rw,nosuid,nodev,noexec,relatime,size=5120k,inode64)
cgroup2 on /sys/fs/cgroup type cgroup2 (rw,nosuid,nodev,noexec,relatime,nsdelegate,memory_recursiveprot)
pstore on /sys/fs/pstore type pstore (rw,nosuid,nodev,noexec,relatime)
efivarfs on /sys/firmware/efi/efivars type efivarfs (rw,nosuid,nodev,noexec,relatime)
bpf on /sys/fs/bpf type bpf (rw,nosuid,nodev,noexec,relatime,mode=700)
systemd-1 on /proc/sys/fs/binfmt_misc type autofs (rw,relatime,fd=29,pgrp=1,timeout=0,minproto=5,maxproto=5,direct,pipe_ino=22834)
hugetlbfs on /dev/hugepages type hugetlbfs (rw,relatime,pagesize=2M)
mqueue on /dev/mqueue type mqueue (rw,nosuid,nodev,noexec,relatime)
debugfs on /sys/kernel/debug type debugfs (rw,nosuid,nodev,noexec,relatime)
tracefs on /sys/kernel/tracing type tracefs (rw,nosuid,nodev,noexec,relatime)
fusectl on /sys/fs/fuse/connections type fusectl (rw,nosuid,nodev,noexec,relatime)
configfs on /sys/kernel/config type configfs (rw,nosuid,nodev,noexec,relatime)
none on /run/credentials/systemd-sysusers.service type ramfs (ro,nosuid,nodev,noexec,relatime,mode=700)
/dev/sda3 on /mnt/C type fuseblk (rw,relatime,user_id=0,group_id=0,allow_other,blksize=4096)
/dev/sda1 on /boot/efi type vfat (rw,relatime,fmask=0007,dmask=0007,allow_utime=0020,codepage=437,iocharset=iso8859-1,shortname=mixed,errors=remount-ro)
/dev/sdb2 on /data type ext4 (rw,relatime)
/dev/sdb1 on /mnt/D type fuseblk (rw,relatime,user_id=0,group_id=0,allow_other,blksize=4096)
binfmt_misc on /proc/sys/fs/binfmt_misc type binfmt_misc (rw,nosuid,nodev,noexec,relatime)
tmpfs on /run/user/1000 type tmpfs (rw,nosuid,nodev,relatime,size=794916k,nr_inodes=198729,mode=700,uid=1000,gid=1000,inode64)
portal on /run/user/1000/doc type fuse.portal (rw,nosuid,nodev,relatime,user_id=1000,group_id=1000)
```
I've mentioned ntfs-3g since I stumbled on [this answer](https://serverfault.com/a/304356) before and thought that's still the case for Ubuntu based distros. I'm not aware whether this has changed.

**Comment by karelzak** [comment]
It seems like no fstrim issue. Closing.

**Comment by lczerner** [comment]
@ahmetsait 

> I've mentioned ntfs-3g since I stumbled on [this answer](https://serverfault.com/a/304356) before and thought that's still the case for Ubuntu based distros. I'm not aware whether this has changed.

The answer in the link is misleading.  ntfs-3g, ntfs, and ntfs3 are all different things. Nevertheless, from the mount output you're using the fuse driver and as Karel mentioned it does not look like a fstrim issue. You should file a bug with the ntfs-3g project instead.

-Lukas

**Comment by ahmetsait** [comment]
Thanks for the comments, I updated the bug report here: https://bugs.launchpad.net/ubuntu/+source/ntfs-3g/+bug/1823721

If there is anything else you guys would recommend me to do please mention.
